wavefront_python_sdk/direct.py

```diff
@@ -0,0 +1,247 @@
+    WAVEFRONT_METRIC_FORMAT = 'wavefront'
+    WAVEFRONT_HISTOGRAM_FORMAT = 'histogram'
+    WAVEFRONT_TRACING_SPAN_FORMAT = 'trace'
+
+    def __init__(self, server, token, max_queue_size=50000, batch_size=10000,
+                 flush_interval_seconds=5):
+        """
+        Constructor of Direct Client
+        @param server: Server address, Example: https://INSTANCE.wavefront.com
+        @type server: str
+        @param token: Token with Direct Data Ingestion permission granted
+        @type token: str
+        @param max_queue_size:
+        @type max_queue_size: int
+        Max Queue Size, size of internal data buffer for each data type.
+        50000 by default
+        @param batch_size:
+        @type batch_size: int
+        Batch Size, amount of data sent by one api call, 10000 by default
+        @param flush_interval_seconds: Interval flush time, 5 seconds by default
+        @type flush_interval_seconds: int
+        """
+        ConnectionHandler.__init__(self)
+        self.server = server
+        self._token = token
+        self._max_queue_size = max_queue_size
+        self._batch_size = batch_size
+        self._flush_interval_seconds = flush_interval_seconds
+        self._default_source = "wavefrontDirectSender"
+        self._metrics_buffer = queue.Queue(max_queue_size)
+        self._histograms_buffer = queue.Queue(max_queue_size)
+        self._tracing_spans_buffer = queue.Queue(max_queue_size)
+        self._headers = {'Content-Type': 'application/octet-stream',
+                         'Content-Encoding': 'gzip',
+                         'Authorization': 'Bearer ' + token}
+        self._flush()
+
+    def _report(self, points, data_format):
+        """
+        One api call sending one given string data
+        @param points: List of data in string format, concat by "\n"
+        @type points: str
+        @param data_format: Type of data to be sent
+        @type data_format: str
+        """
+        try:
+            params = {'f': data_format}
+            compressed_data = gzip_compress(points.encode())
+            response = requests.post(self.server + '/report', params=params,
+                                     headers=self._headers,
+                                     data=compressed_data)
+            response.raise_for_status()
+        except Exception as e:
+            self.increment_failure_count()
+            raise e
+
+    def _batch_report(self, batch_line_data, data_format):
+        """
+        One api call sending one given list of data
+        @param batch_line_data: List of data to be sent
+        @type batch_line_data: list
+        @param data_format: Type of data to be sent
+        @type data_format: str
+        """
+        # Split data into chunks, each with the size of given batch_size
+        for batch in chunks(batch_line_data, self._batch_size):
+            # report once per batch
+            self._report('\n'.join('{0}'.format(line) for line in batch) + "\n",
```

Comment:
`'{0}'.format(line)` is useless if all items in `batch` are strings, or can be replaced with `str(line)`

Reply:
This would be `'\n'.join(batch) + '\n'`